A45-retry-stats.md

```diff
@@ -0,0 +1,126 @@
+
+
+### Metrics to Expose
+
+gRPC will treat each retry attempt or hedged RPC as a distinct RPC with regards to the current set of [metrics]((https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/gRPC.md#measures)).
+
+We add the following per-overall-client-call metrics to OpenCensus:
+
+#### Measures
+
+| Measure name                                   | Unit | Description                                                                                                                                                    |
+| ---------------------------------------------- | ---- | -------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| grpc.io/client/retries\_per\_call              | 1    | Number of retry or hedging attempts excluding transparent retries made during the client call. The original attempt is not counted as a retry/hedging attempt. |
+| grpc.io/client/transparent\_retries\_per\_call | 1    | Number of transparent retries made during the client call.                                                                                                     |
+| grpc.io/client/retry\_delay\_per\_call         | ms   | Total time of delay while there is no active attempt during the client call.           
+
+#### Views
+
+| View name                                      | Measure                         | Aggregation  | tags                 |
+| ---------------------------------------------- | ------------------------------- | ------------ | -------------------- |
+| grpc.io/client/retries\_per\_call              | retries\_per\_call              | distribution | grpc\_client\_method |
+| grpc.io/client/retries                         | retries\_per\_call              | sum          | grpc\_client\_method |
+| grpc.io/client/transparent\_retries\_per\_call | transparent\_retries\_per\_call | distribution | grpc\_client\_method |
+| grpc.io/client/transparent\_retries            | transparent\_retries\_per\_call | sum          | grpc\_client\_method |
+| grpc.io/client/retry\_delay\_per\_call         | retry\_delay\_per\_call         | distribution | grpc\_client\_method |
+
+
+### Tracing
+
+If retry is not enabled on the client channel, the gRPC tracing module should behave the same as it does today. If retry is enabled, the gRPC tracing module should create a tracing span for each call from the perspective of the application, and create a child span for each individual call attempt, including transparent retry attempts. The parent span is named in the form of `Sent.<service name>.<method name>`, which is the same as the case when retry is not enabled. The span for each individual call attempt is named in the form of `Attempt.<service name>.<method name>`, and is with the following additional span attributes:
+
+      `previous-rpc-attempts` : number of preceding attempts,
```

Comment:
Is `previous-rpc-attempts` incremented due to a transparent retry? I see it stated later "Transparent retry should also be treated as a distinct RPC with regards to the per-attempt metrics" but not stated here. If so, does that mix well with hedging?

Reply:
Good question. @markdroth , does the `grpc-previous-rpc-attempts` header in the retry gRFC increment on a transparent retry? IIRC it does not increment on transparent retry.

Reply:
Java currently is not incrementing this count for transparent retries.

Reply:
Defined the attribute excluding transparent retry.